src/Features/Core/Portable/FindUsages/DefinitionItem.DocumentLocationDefinitionItem.cs

```diff
@@ -11,25 +17,112 @@ internal partial class DefinitionItem
+                }
+
+                if (this.Properties.TryGetValue(MetadataSymbolKey, out var symbolKey))
+                {
+                    return CanNavigateToMetadataSymbol(symbolKey);
+                }
+
+                return SourceSpans[0].CanNavigateTo();
+            }
+
+            public override bool TryNavigateTo()
+            {
+                if (this.Properties.ContainsKey(NonNavigable))
+                {
+                    return false;
+                }
+
+                if (this.Properties.TryGetValue(MetadataSymbolKey, out var symbolKey))
+                {
+                    return TryNavigateToMetadataSymbol(symbolKey);
+                }
+
+                return SourceSpans[0].TryNavigateTo();
+            }
+
+            private bool CanNavigateToMetadataSymbol(string symbolKey)
+                => TryNavigateToMetadataSymbol(symbolKey, (symbol, project, service) => true);
+
+            private bool TryNavigateToMetadataSymbol(string symbolKey)
+                => TryNavigateToMetadataSymbol(symbolKey, (symbol, project, service) =>
+                    service.TryNavigateToSymbol(
+                        symbol, project, project.Solution.Options.WithChangedOption(NavigationOptions.PreferProvisionalTab, true)));
+
+            private bool TryNavigateToMetadataSymbol(string symbolKey, Func<ISymbol, Project, ISymbolNavigationService, bool> action)
+            {
+                var projectAndSymbol = ResolveSymbolInCurrentSolution(symbolKey);
+                if (projectAndSymbol == null)
+                {
+                    return false;
+                }
+
+                var project = projectAndSymbol?.project;
+                var symbol = projectAndSymbol?.symbol;
+                if (symbol == null || project == null)
+                {
+                    return false;
+                }
+
+                if (symbol.Kind == SymbolKind.Namespace)
+                {
+                    return false;
+                }
+
+                var navigationService = _workspaceOpt.Services.GetService<ISymbolNavigationService>();
+                return action(symbol, project, navigationService);
+            }
+
+            private (Project project, ISymbol symbol)? ResolveSymbolInCurrentSolution(string symbolKey)
+            {
+                if (!this.Properties.TryGetValue(MetadataAssemblyIdentityDisplayName, out var identityDisplayName) ||
+                    !AssemblyIdentity.TryParseDisplayName(identityDisplayName, out var identity))
+                {
+                    return null;
+                }
+
+                var project = _workspaceOpt.CurrentSolution
+                    .ProjectsWithReferenceToAssembly(identity)
+                    .FirstOrDefault();
+
+                if (project == null)
+                {
+                    return null;
+                }
+
+                var compilation = project.GetCompilationAsync(CancellationToken.None)
```

Comment:
💡 Consider making the enclosing method asynchronous and pushing the `WaitAndGetResult` as far up the call chain as possible.